### PR TITLE
DEVX-1520: Add tutorial for KStreams Cogrouping

### LINKED ISSUE #199: Using Kafka Streams Co-Grouping
Tutorial on how to use Kafka Streams [Cogroup functionality](https://cwiki.apache.org/confluence/display/KAFKA/KIP-150+-+Kafka-Streams+Cogroup)

### PR DESCRIPTION
**Pull request by bbejeck**
Fixes #199

### CODE REVIEW

**Comment by bbejeck** [comment]
Verified locally by going through tutorial steps and running test-harness locally.

**Comment by bbejeck** [comment]
ping @colinhicks @rspurgeon @gAmUssA for review

File: _includes/tutorials/cogrouping-streams/kstreams/code/src/main/java/io/confluent/developer/CogroupingStreams.java [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+public class CogroupingStreams {
+
+
+	public Properties buildStreamsProperties(Properties envProps) {
+        Properties props = new Properties();
+
+        props.put(StreamsConfig.APPLICATION_ID_CONFIG, envProps.getProperty("application.id"));
+        props.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, envProps.getProperty("bootstrap.servers"));
+        props.put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath());
+        props.put(AbstractKafkaAvroSerDeConfig.SCHEMA_REGISTRY_URL_CONFIG, envProps.getProperty("schema.registry.url"));
```
Comment (gAmUssA) [inline _includes/tutorials/cogrouping-streams/kstreams/code/src/main/java/io/confluent/developer/CogroupingStreams.java:47]:
```suggestion
        props.put(AbstractKafkaSchemaSerDeConfig.SCHEMA_REGISTRY_URL_CONFIG, envProps.getProperty("schema.registry.url"));
```

Comment (gAmUssA) [reply _includes/tutorials/cogrouping-streams/kstreams/code/src/main/java/io/confluent/developer/CogroupingStreams.java:47]: `AbstractKafkaAvroSerDeConfig` is deprecated in 5.5

File: _includes/tutorials/cogrouping-streams/kstreams/code/src/main/java/io/confluent/developer/CogroupingStreams.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
+package io.confluent.developer;
+
+
+import io.confluent.common.utils.TestUtils;
+import io.confluent.developer.avro.LoginEvent;
+import io.confluent.developer.avro.LoginRollup;
+import io.confluent.kafka.serializers.AbstractKafkaAvroSerDeConfig;
```
Comment (gAmUssA) [inline _includes/tutorials/cogrouping-streams/kstreams/code/src/main/java/io/confluent/developer/CogroupingStreams.java:7]:
```suggestion
```

File: _includes/tutorials/cogrouping-streams/kstreams/code/src/main/java/io/confluent/developer/CogroupingStreams.java [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+import io.confluent.kafka.serializers.AbstractKafkaSchemaSerDeConfig;
+import io.confluent.kafka.serializers.KafkaAvroDeserializer;
+import io.confluent.kafka.serializers.KafkaAvroSerializer;
+import io.confluent.kafka.streams.serdes.avro.PrimitiveAvroSerde;
```
Comment (gAmUssA) [inline _includes/tutorials/cogrouping-streams/kstreams/code/src/main/java/io/confluent/developer/CogroupingStreams.java:11]:
```suggestion
```

**Comment by bbejeck** [comment]
@gAmUssA all changes addressed

ping @colinhicks for second review before merging

File: _includes/tutorials/cogrouping-streams/kstreams/markup/dev/make-aggregator.adoc [hunk 4]
```diff
@@ -0,0 +1,27 @@
+////
+In this file you describe the Kafka streams topology, and should cover the main points of the tutorial.
+The text assumes a method buildTopology exists and constructs the Kafka Streams application.  Feel free to modify the text below to suit your needs.
+////
+
+The `Aggregator` you saw in the previous step, constructs a map keeping a count of user logins per user, per application, so a map of maps.  Here you'll see the core logic of the `LoginAggregator`, and the code is straightforward, as you can see below.
```
Comment (colinhicks) [inline _includes/tutorials/cogrouping-streams/kstreams/markup/dev/make-aggregator.adoc:6]:
```suggestion
The `Aggregator` you saw in the previous step constructs a map of maps: the count of logins per user, per application.  Below is the core logic of the `LoginAggregator`.
```

File: _includes/tutorials/cogrouping-streams/kstreams/markup/dev/make-aggregator.adoc [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+Each call to `Aggregator.apply` retrieves the user login map for the given application id (or creating one if it doesn't exist).  From there, the `Aggregator` increments the login count for the given user.
```
Comment (colinhicks) [inline _includes/tutorials/cogrouping-streams/kstreams/markup/dev/make-aggregator.adoc:8]:
```suggestion
Each call to `Aggregator.apply` retrieves the user login map for the given application id (or creates one if it doesn't exist).  From there, the `Aggregator` increments the login count for the given user.
```

File: _includes/tutorials/cogrouping-streams/kstreams/markup/dev/make-avro-dir.adoc [hunk 6]
```diff
@@ -0,0 +1,8 @@
+
+This tutorial uses 4 streams.  The three input streams have a record type of `LoginEvent` used to represent a user logging into an application.  The fourth stream is an ouput stream that writes a `LoginRollup` object out to a topic.  In the next steps you'll create the avro schemas for these objects.
```
Comment (colinhicks) [inline _includes/tutorials/cogrouping-streams/kstreams/markup/dev/make-avro-dir.adoc:2]:
```suggestion
This tutorial uses 4 streams.  The three input streams have a record type of `LoginEvent` used to represent a user logging into an application.  The fourth stream is an output stream that writes a `LoginRollup` object out to a topic.  In the next steps you'll create the Avro schemas for these objects.
```

File: _includes/tutorials/cogrouping-streams/kstreams/markup/dev/make-topology.adoc [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+            .aggregate(() -&gt; new LoginRollup(new HashMap&lt;>()), Materialized.with(Serdes.String(), loginRollupSerde))
+            .toStream().to(totalResultOutputTopic, Produced.with(stringSerde, loginRollupSerde));
+    </code>
+</pre>
+++++
+
+You're using the cogrouping functionality here to get an overall grouping of logins per application.  Kafka Streams creates this total grouping by using an `Aggregator` who knows how to extract records from each grouped stream.  Your `Aggregator` instance here knows how to correctly combine each `LoginEvent` into the larger `LoginRollup` object.  You'll learn more about `Aggregator` in the next step.
+
+Next, you have three input streams (`appOneStream`, `appTwoStream`, and `appThreeStream`).  You need the intermediate object `KGroupedStream`  so you execute the
+`groupByKey()` method on each stream.  For this tutorial, we have assumed the incoming records already have keys.  Otherwise, you'll need to use a key selecting method (`selectKey()`, `map()`), or `groupBy()` to successfully group by key.
```
Comment (colinhicks) [inline _includes/tutorials/cogrouping-streams/kstreams/markup/dev/make-topology.adoc:29]:
```suggestion
Next, you have three input streams: `appOneStream`, `appTwoStream`, and `appThreeStream`.  You need the intermediate object `KGroupedStream`, so you execute the `groupByKey()` method on each stream.  For this tutorial, we have assumed the incoming records already have keys.  In cases where records lack keys, you need to use a key-selecting method (`selectKey()`, `map()`, or `groupBy()`) to successfully group by key.
```

File: _includes/tutorials/cogrouping-streams/kstreams/markup/dev/make-topology.adoc [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+Now with your `KGroupedStream` objects, you start creating your larger aggregate by calling `KGroupedStream.cogroup()` on the first stream, using your `Aggregator`.  This first step returns a `CogroupedKStream` instance.  Then for each remaining `KGroupedStream`, you execute `CogroupedKSteam.cogroup()` using one of the `KGroupedStream` instances and the `Aggregator` you created previously.  You repeat this sequence of calls for all of the `KGroupedStream` objects you want to combine into an overall aggregate.
+
+For more background on cogrouping functionality in stream you can https://cwiki.apache.org/confluence/display/KAFKA/KIP-150+-+Kafka-Streams+Cogroup[read the KIP proposal].
```
Comment (colinhicks) [inline _includes/tutorials/cogrouping-streams/kstreams/markup/dev/make-topology.adoc:33]:
```suggestion
For more background on cogrouping functionality in stream you can https://cwiki.apache.org/confluence/display/KAFKA/KIP-150+-+Kafka-Streams+Cogroup[read the KIP-150 proposal].
```

File: _data/tutorials.yaml [hunk 9]
```diff
@@ -257,3 +257,14 @@ naming-changelog-repartition-topics:
     ksql: disabled
     kstreams: enabled
     kafka: disabled
+
+cogrouping-streams:
+  title: "How to combine stream aggregates together in a single larger object"
+  meta-description: "How to combine stream aggregates together in a single larger object"
+  slug: "/cogrouping-streams"
+  problem: "you have multiple streams with an aggregate value that you want to combine into a single object"
```
Comment (colinhicks) [inline _data/tutorials.yaml:265]:
```suggestion
  problem: "you have multiple streams, each with an aggregate value like `count`,  that you want to combine into a single result"
```

File: _data/tutorials.yaml [hunk 10]
```diff
@@ -257,3 +257,14 @@ naming-changelog-repartition-topics:
...
+  introduction: "You have mulitple streams with an aggregate value but you want to group them all together - cogrouping.  In this tutorial we'll cover how to use the KStreams CoGrouping functionality to merge mulitple aggregates into a single aggregate object"
```
Comment (colinhicks) [inline _data/tutorials.yaml:266]:
```suggestion
  introduction: "You want to compute the count of user login events per application in your system, grouping the individual result from each source stream into one aggregated object.  In this tutorial we'll cover how to use the Kafka Streams Cogrouping functionality to accomplish this task with clear, performant code"
```

File: .gitignore [hunk 11]
```diff
@@ -9,7 +9,7 @@ build/
 .jekyll-metadata
 node_modules/
 out/
-*/outputs/*
+**/tutorial-steps/dev/outputs/
```
[the remark this answers is not in the document]
Comment (bbejeck) [reply .gitignore:12]: That's a good point, I'll add a separate .gitignore file for the tutorial.

**Comment by bbejeck** [comment]
@colinhicks all comments addressed

**Pull request merged**